### ISSUE TITLE
[Mining] PowBlockAssembler race condition

### ISSUE THREAD

**Issue by Aprogiena** [stratisproject/StratisBitcoinFullNode#500]
POS Miner crashed with
```
 PosMining.Mine threw an unhandled exception: System.ArgumentException: The previous block has not the expected hash
   at Stratis.Bitcoin.Features.Miner.PosMinting.<>c__DisplayClass35_0.<Mine>b__0(CancellationToken token) in C:\StratisBitcoinFullNode\Stratis.Bitcoin.Features.Miner\PosMinting.cs:line 206
   at Stratis.Bitcoin.Utilities.AsyncLoop.<>c__DisplayClass12_0.<<StartAsync>b__0>d.MoveNext() in C:\StratisBitcoinFullNode\Stratis.Bitcoin\Utilities\AsyncLoop.cs:line 129
```

The origin of this exception is:
`PosMinting.GenerateBlocks` -> `PosBlockAssembler.CreateNewBlock` -> `PowBlockAssembler.CreateNewBlock` ->  `PosBlockAssembler.UpdateHeaders` -> `PowBlockAssembler.UpdateHeaders`

here we call 
```
            this.pblock.Header.HashPrevBlock = this.pindexPrev.HashBlock;
            this.pblock.Header.UpdateTime(this.dateTimeProvider.GetTimeOffset(), this.network, this.chain.Tip);
            this.pblock.Header.Bits = this.pblock.Header.GetWorkRequired(this.network, this.chain.Tip);
```

which already by itself seems wrong because those two lines potentially use different chain Tip. The problem is then raised from `GetWorkRequired`, which leads to

```
return new ChainedBlock(this, null, prev).GetWorkRequired(consensus);
```
and the constructor has this check
```
				if(previous.HashBlock != header.HashPrevBlock)
					throw new ArgumentException("The previous block has not the expected hash");
```

Solution is probably to use `pindexPrev` instead of `this.chain.Tip` in both lines.

**Comment by Aprogiena** [comment]
![](https://github.trello.services/images/mini-trello-icon.png) [[Mining] PowBlockAssembler race condition](https://trello.com/c/ozmGxAny/262-mining-powblockassembler-race-condition)

**Comment by Aprogiena** [comment]
However, a better fix would probably be to pass chain tip as an argument to `CreateNewBlock` or maybe even the constructor of `PowBlockAssembler`.

Any thoughts @dangershony ?

**Comment by Aprogiena** [comment]
Same for `PosBlockAssembler` and its `UpdateHeaders`, there is consensusLoop.Tip used, which is better than chain.Tip, but still vulnerable

**Comment by dangershony** [comment]
Can I close this?

**Comment by Aprogiena** [comment]
Why would you? No effort was done to fix the problem, it is currently the number one problem that is crashing the POS mining process.

**Comment by dangershony** [comment]
Ah I thought you fixed it with the #487 commit, so this still causes the node to crash?

**Comment by Aprogiena** [comment]
No, this is separate issue.
